src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/AbstractConverter.cs

```diff
@@ -93,5 +114,166 @@ private static QueryClauseSyntax CreateQueryClause(ExtendedSyntaxNode node)
+            var node = ForEachInfo.ConvertingExtendedNodes[extendedNodeIndex];
+            switch (node.Node.Kind())
+            {
+                // Nested ForEach statement is converted into a nested Select or SelectMany linq invocation. For example:
+                //
+                // INPUT:
+                //   foreach (var n1 in c1)
+                //      foreach (var n2 in c2)
+                //          ...
+                //
+                // OUTPUT:
+                //   c1.SelectMany(n1 => c2.Select(n2 => ...
+                //
+                case SyntaxKind.ForEachStatement:
+                    hasForEachChild = true;
+                    var foreachStatement = (ForEachStatementSyntax)node.Node;
+                    ++extendedNodeIndex;
+                    return CreateLinqInvocation(
+                        foreachStatement,
+                        receiverForInvocation: foreachStatement.Expression,
+                        selectExpression: selectExpression,
+                        leadingCommentsTrivia: node.ExtraLeadingComments,
+                        trailingCommentsTrivia: node.ExtraTrailingComments,
+                        currentExtendedNodeIndex: ref extendedNodeIndex);
+
+                // Nested If statement is converted into a Where method invocation on the current receiver. For example:
+                //
+                // INPUT:
+                //   foreach (var n1 in c1)
+                //      if (n1 > 0)
+                //          ...
+                //
+                // OUTPUT:
+                //   c1.Where(n1 => n1 > 0).Select(n1 => ...
+                //
+                case SyntaxKind.IfStatement:
+                    var ifStatement = (IfStatementSyntax)node.Node;
+                    var parentForEachStatement = ifStatement.GetAncestor<ForEachStatementSyntax>();
+                    var lambdaParameter = SyntaxFactory.Parameter(SyntaxFactory.Identifier(parentForEachStatement.Identifier.ValueText));
+                    var lambda = SyntaxFactory.SimpleLambdaExpression(
+                        SyntaxFactory.Parameter(
+                            SyntaxFactory.Identifier(parentForEachStatement.Identifier.ValueText)),
+                        ifStatement.Condition.WithCommentsFrom(ifStatement.OpenParenToken, ifStatement.CloseParenToken))
+                        .WithCommentsFrom(ifStatement.IfKeyword.GetAllTrivia().Concat(node.ExtraLeadingComments), node.ExtraTrailingComments);
+
+                    receiver = SyntaxFactory.InvocationExpression(
+                        SyntaxFactory.MemberAccessExpression(
+                            SyntaxKind.SimpleMemberAccessExpression,
+                            receiver.Parenthesize(),
+                            SyntaxFactory.IdentifierName(nameof(Enumerable.Where))),
+                        SyntaxFactory.ArgumentList(SyntaxFactory.SingletonSeparatedList(
+                            SyntaxFactory.Argument(lambda))));
```

Comment:
fwiw, you might have wanted to do these as SyntaxGenerator calls.  There is nothing really language specific here.  By doing that, you'd have an easier time in the future moving this to VB.  def don't have to do this now.  but might be worth a refactoring in the futre to help keep the actual VB PR small.

Reply:
Yes, I agree with you, but I tried to keep the newly added code in sync with all the existing code in this provider. We should definitely move all of this to use the syntax generator/editor instead of the language specific factories.